system/lib/wasmfs/file_table.h

```diff
@@ -58,21 +58,24 @@ class OpenFileState : public std::enable_shared_from_this<OpenFileState> {
   friend FileTable;
 
 public:
-  OpenFileState(private_key, oflags_t flags, std::shared_ptr<File> file)
-    : file(file), flags(flags) {}
+  // Cache directory entries at the moment the directory is opened so that
+  // subsequent getdents calls have a stable view of the contents. Including
+  // files removed after the open and excluding files added after the open is
+  // allowed, and trying to recalculate the directory contents on each getdents
+  // call could lead to missed directory entries if there are concurrent
+  // deletions that effectively move entries back past the current read position
+  // in the open directory.
+  const std::vector<Directory::Entry> dirents;
```

Comment:
Perhaps make this private like the other properties on this class? Or is there a fundamental difference compared to them?

Reply:
This is immutable, so there's no need to access it through the locked handle. I can make it private for consistency if you want, though. We already have a locked handle at the use site, so it shouldn't matter for performance.

Reply:
Ah, immutability... makes sense than as is, thanks.